FSK DAC: 1200bps telemetry over audio (similar to Bel202 encoding)

peabody124 opened the pull request:
This creates a PIOS_COM interace to the DAC output that creates
a 1200bps audio modem. The stream is modulated similarly to V.23
encoding using 1 start and 1 stop bits, 1 parity bit. The two symbols
are centered at 1200 and 2200 Hz.

The data source tested with this is the light telemetry module that was
developed for this purpose of low bandwidth telemetry, previously using
a hardware serial stream to audio modem chip.

peabody124 committed:
PIOS FSKDAC: stub commit of FSK DAC PIOS_COM driver

This will be a pios_com interface that generates
Bell202 encoded data on the DAC DMA channel.

peabody124 committed:
PIOS FSKDAC: much closer to working

peabody124 committed:
PIOS FSKDAC: create include file

peabody124 committed:
PIOS FSKDAC: more cleanup, compiles now

peabody124 committed:
Sparky2: enable FSKDAC

peabody124 committed:
PIOS FSKDAC: Sine wave output works now

peabody124 committed:
PIOS FSKDAC: tone generation working

peabody124 committed:
PIOS FSKDAC: ChiBios IRQ epilogue/prologue

peabody124 committed:
PIOS FSKDAC: use defined register address

peabody124 committed:
PIOS FSKDAC: bump copyright

peabody124 committed:
PIOS FSKDAC: update the symbols to fit in VTX range

peabody124 committed:
PIOS FSKDAC: use PIOS_DAC_COMMON and small cleanup

peabody124 committed:
PIOS FSKDAC: get mark and space correct

peabody124 committed:
Sparky2: add PIOS COM layer to FSK DAC

peabody124 committed:
PIOS FSKDAC: now properly responds to ComBridge

peabody124 committed:
Sparky2: FSK DAC sends light telem when enabled

peabody124 committed:
PIOS FSKDAC: minimum phase between symbols

Because MARK is 1200Hz and SPACE is 2200Hz at 1200bps
there will not be an integer number of cycles for a
space. Thus we have to work out the phase we end each
symbol at and make sure to start the next one at that
phase.

This also drops the code that sends a mid-voltage
between symbols and it constantly outputs MARK when
awaiting a new byte.

peabody124 committed:
PIOS FSKDAC: remove debug code

peabody124 committed:
PIOS FSKDAC: only access TX CB from IRQ

peabody124 committed:
PIOS FSKDAC: use stdperiph for configuring next symbol

peabody124 committed:
LightTelemetry: increase stack size and send nonblocking

peabody124 committed:
PIOS DAC Common: lower priority of DMQ IRQ

peabody124 committed:
PIOS FSKDAC: stop bit is also SPACE

peabody124 committed:
PIOS FSKDAC: add parity bit

peabody124 committed:
PIOS DAC COMMON: support the DMA HT flag

This is required for FSKDAC since we use the middle of a
DMA to schedule what the next symbol should be, rather than
waiting to the end.

peabody124 committed:
PIOS FSKDAC: update symbols during DMA half transfer

This makes sure we are prepared to send the next symbol when
the buffer swaps.

peabody124 committed:
PIOS FSKDAC: smaller amplitude with less distortion

peabody124 committed:
PIOS FSKDAC: do not disable DMA IRQ between bytes

For some reason, disabling the DMA HT flag, which should
reduce the CPU utilization, ultimately results in some of
the start symbols being dropped. There doesn't appear to
be a clear software bug, so perhaps this is some kind of
hardware erata when rapidly toggling IRQ enables.

peabody124 committed:
PIOS FSKDAC: bump copyright

peabody124 committed:
Sparky2: FSKDAC buffer should be 19 bytes

This is enough to hold the largest packet but too
small to hold two packets. This prevents them stacking
up in the COM buffer and introducing latency.

peabody124 committed:
LightTelemetry: support builtin module

peabody124 commented:
Here is a demo of it working with real time decoding and visualization into ROS
https://www.instagram.com/p/BCJYDuglqrb/

kubark42 commented:
Neat!!!

kubark42 commented on flight/PiOS/STM32F4xx/pios_fskdac.c:
```diff
@@ -0,0 +1,352 @@
+};
+
+enum pios_fskdac_dev_magic {
+	PIOS_FSKDAC_DEV_MAGIC = 0x1453834A,
```
Any method for making the magic number? Or just more keyboard banging? ;)

peabody124 replied:
Sorry this predated your suggestion so monkey meet keyboard

kubark42 replied:
Heheh, no prob.

kubark42 commented on flight/PiOS/STM32F4xx/pios_fskdac.c:
```diff
@@ -0,0 +1,352 @@
+};
+
+struct pios_fskdac_dev {
+	enum pios_fskdac_dev_magic     magic;
```
Extra whitespace.

kubark42 commented on flight/PiOS/STM32F4xx/pios_fskdac.c:
```diff
@@ -0,0 +1,352 @@
+
+	PIOS_DAC_COMMON_Init(PIOS_FSKDAC_DMA_irq_cb);
+
+	TIM_SetAutoreload(TIM6, PIOS_PERIPHERAL_APB1_CLOCK / DAC_SBS);
```
Will this wind up changing frequency on chips with different APB1 clocks?

peabody124 replied:
not if they correctly define that frequency, this should make it do the right thing

kubark42 commented on flight/PiOS/STM32F4xx/pios_fskdac.c:
```diff
@@ -0,0 +1,352 @@
+
+static bool PIOS_FSKDAC_validate(struct pios_fskdac_dev * fskdac_dev)
+{
+	return (fskdac_dev->magic == PIOS_FSKDAC_DEV_MAGIC);
```
Thanks for putting this in `()`.

peabody124 committed:
PIOS CAN: support vertical rate msg

This provides the smoothed altitude and climb rate

peabody124 committed:
LightTelemetry: send smoothed altitude when available

For such a low bandwidth link we should send the smoothed
altitude rather than raw baro samples to reduce noise.

peabody124 committed:
Sparky2: set LTM to 1200bps when FSK

kubark42 commented on flight/Modules/UAVOLighttelemetryBridge/UAVOLighttelemetryBridge.c:
```diff
@@ -44,21 +44,23 @@
-#include "accels.h"
 #include "manualcontrolcommand.h"
-#include "flightstatus.h"
+#include "positionactual.h"
```
Alphabetization, my favorite!

kubark42 commented on flight/Modules/UAVOLighttelemetryBridge/UAVOLighttelemetryBridge.c:
```diff
@@ -44,21 +44,23 @@
  */
 #include "openpilot.h"
 #include "modulesettings.h"
```
This straggler could probably move down into the UAVObject includes.

kubark42 commented on flight/targets/sparky2/fw/pios_board.c:
```diff
@@ -696,8 +699,20 @@ void PIOS_Board_Init(void) {
+		if (PIOS_COM_Init(&fskdac_com_id, &pios_fskdac_com_driver, fskdac_id,
+		                  NULL, 0,
+		                  tx_buffer, PIOS_COM_FSKDAC_BUF_LEN))
+			panic(6);
```
At some point, we should make enums for the `panic()` so that we can guarantee "sameness" across boards.

tracernz replied:
I have that in a branch (slight refactor to existing pios_hal_panic + enums). Just need to finish putting into all the targets.

kubark42 replied:
Cool. Looking forward to it.

On a related note, I've been thinking about going through the entire firmware codebase and fixing all the errors returned by all the functions. Maybe just bundle them all into one giant enum. I think it would help us do better error handling if we had names for them, instead of just `-1`, `0`, etc...

tracernz replied:
Wip: https://github.com/tracernz/dronin/commits/keep-calm-and-carry-on

kubark42 commented:
Good stuff. I'm trusting that the modem math works, I haven't vetted it very closely. The only remaining question before merging is whether you'd like to move the `#include modulesettings.h` line down with the other UAVObjects.

peabody124 committed:
FSKDAC: feedback from review

peabody124 commented:
Yeah the math works. If you have a clever thought about converting the phase alignment calculation
https://github.com/TauLabs/TauLabs/pull/2202/files#diff-66ff64856df7b2196f418ca7a9ebd111R221
to integer math I'd be happy to hear it, although would probably defer to another PR for improvement.

peabody124 commented:
PS comments from review should be addressed now.

kubark42 commented:
> If you have a clever thought about converting the phase alignment calculation
> https://github.com/TauLabs/TauLabs/pull/2202/files#diff-66ff64856df7b2196f418ca7a9ebd111R221
> to integer math I'd be happy to hear it

Nothing instantly occurs to me, but I'm not sure the `fmodf()` function is that expensive since it is basically an FPU division followed by an FPU multiplication and subtraction. How much precision do you need?

mlyle commented:
There's a bit more to it than that-- you burn significant precision with those operations.

https://github.com/eblot/newlib/blob/master/newlib/libm/math/ef_fmod.c

peabody124 commented:
@kubark42 not much precision since it just determines an integer offset into the next buffer

kubark42 commented on flight/PiOS/STM32F4xx/pios_fskdac.c:
```diff
@@ -0,0 +1,352 @@
+ * Schedule the next bit to be written by setting the address of the
+ * double buffer
+ */
+static void pios_fskdac_set_symbol(struct pios_fskdac_dev * fskdac_dev, uint8_t sym)
```
I just looked at this section a little more. If you have a moment, a future PR which expanded the theoretic justification for this section would be most welcome. Perhaps just links to the documents you used to develop this would be enough to maintain it.

peabody124 replied:
Sure. Probably best for me to just explain here since I created this based on just tying to achieve continuous phase while transitioning between symbols.

Unfortunately good sources on Bell 202 are lacking, so I can't be sure this is fully compatible with hardware implementations.  That would be reflected in the state system below as it reflects more on how you map a byte into a sequence of mark/space symbols.

peabody124 commented:
BTW @kubark42 check out https://www.instagram.com/p/BCRgrnjFquF/ so now any FPV recording has a data log (slow).

peabody124 commented:
Also @KipK not sure if you are still active but thought you might like this. I'd love to know if this is actually compatible with the chip you use in ghettostation.

peabody124 commented:
![screen shot 2016-02-27 at 12 07 27 pm](https://cloud.githubusercontent.com/assets/618346/13374501/f1f33d72-dd4b-11e5-917b-6ea72db18e64.png)

kubark42 commented:
That's pretty neat stuff. Are you exporting it to KML?

peabody124 commented:
@kubark42 yeah. And there is a neat gx:Track type that Google earth supports animates the flight path too, which is kind of fun

kubark42 commented on flight/PiOS/STM32F4xx/pios_fskdac.c:
```diff
@@ -0,0 +1,352 @@
+	}
+	
+#if defined(PIOS_INCLUDE_FREERTOS)
+	portEND_SWITCHING_ISR((rx_need_yield || tx_need_yield) ? pdTRUE : pdFALSE);
```
When compiling with FreeRTOS, this fails with unknown variables.
